apps/btc_family/btc_txn_helpers.c

```diff
@@ -0,0 +1,248 @@
+  weight = 4 * weight;    // As per standard non segwit transaction size is
+                          // multiplied by 4
+
+  if (segwit_count > 0) {
+    weight += 2;    // Segwit headers
+    weight +=
+        (106 * segwit_count);    // Adding sizes of all witnesses for all inputs
+  }
+
+  return weight;
+}
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+
+int btc_verify_input_utxo(const uint8_t *raw_txn,
+                          const uint32_t size,
+                          const btc_txn_input_t *input) {
+  if (NULL == input || NULL == raw_txn || 0 == size) {
+    return -1;
+  }
+
+  uint8_t hash[SHA256_DIGEST_LENGTH] = {0};
+  int32_t offset = 0;
+  int32_t count = 0;
+  int32_t value_offset = 0;
+  int32_t start_offset = 0;
+  uint8_t txn_data[size];
+
+  // ignore network version (4-bytes), skip marker & flag (in segwit)
+  offset += (raw_txn[4] == 0 ? 6 : 4);
+  start_offset = offset;
+  // remember the number of input UTXOs in the raw_txn
+  count = raw_txn[offset++];
+
+  for (int32_t input_index = 0; input_index < count; input_index++) {
+    offset += 36;
+    offset += (raw_txn[offset] + 1 + 4);
+  }
+  // remember the number of output UTXOs in the raw_txn
+  count = raw_txn[offset++];
+  for (int32_t output_index = 0; output_index < count; output_index++) {
+    if (output_index == input->prev_output_index) {
+      // only check the specified output index as we are looking for an exact
+      // match; remember the location of value in the raw_txn & compare later
+      value_offset = offset;
+
+      // NOTE: do not break here, the 'offset' should traverse till the end
+    }
+    offset += 8;
+    offset += (raw_txn[offset] + 1);
+  }
+
+  if (value_offset == 0) {
+    return 1;
+  }
+
+  // network version (first 4 bytes)
+  memcpy(txn_data, raw_txn, 4);
+  // txin and txout (skip marker & flag)
+  memcpy(txn_data + 4, raw_txn + start_offset, offset - start_offset);
+  // locktime (last 4 bytes)
+  memcpy(txn_data + offset - start_offset + 4, raw_txn + size - 4, 4);
+  sha256_Raw(txn_data, offset - start_offset + 4 + 4, hash);
+  sha256_Raw(hash, sizeof(hash), hash);
+  // verify UTXO txn hash
+  if (memcmp(hash, input->prev_txn_hash, sizeof(input->prev_txn_hash)) != 0) {
+    return 2;
+  }
+  if (U64_READ_LE_ARRAY(raw_txn + value_offset) != input->value) {
+    return 3;
+  }
+  return 0;
+}
+
+uint64_t get_transaction_fee_threshold(const btc_txn_context_t *txn_ctx) {
+  return (g_app->max_fee / 1000) * (get_transaction_weight(txn_ctx) / 4);
+}
+
+uint64_t btc_get_txn_fee(const btc_txn_context_t *txn_ctx) {
+  uint64_t result = 0;
+
+  for (int idx = 0; idx < txn_ctx->metadata.input_count; idx++) {
+    result += txn_ctx->inputs[idx].value;
+  }
+
+  for (int idx = 0; idx < txn_ctx->metadata.output_count; idx++) {
+    result -= txn_ctx->outputs[idx].value;
+  }
+
+  return result;
```

Comment:
We need to ensure that input value is always greater than or equal to output value.. otherwise this arithmetic could lead to negative values being interpreted as very large uint values..

Reply:
changed